### PR TITLE
Farmer Enhancement

### LINKED ISSUE #231: Make Farmer behave in a more optimal fasion
The Farmer should farm the fields in a manner that resembles someone of at least minor intelligence

### LINKED ISSUE #232: Pam's HarvestCraft compatibility
Farmer compatibility with Pam's HarvestCraft

### PR DESCRIPTION
**Pull request by Dadadah**
Closes #231 

# Changes proposed in this pull request:
- ~~Pam's HarvestCraft Compatibility~~ Make farmers harvest like Pam's
- Farmer optimizations
- Farmer smartifyers

Review please @Minecolonies/maintainers

### CODE REVIEW

**Comment by Dadadah** [comment]
Fixed my git config, it now shows my email, sorry. I have signed CLA

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 1]
```diff
@@ -568,4 +588,54 @@ public EntityCitizen getCitizen()
     {
         return worker;
     }
+
+    private boolean harvestCrop(final BlockPos position)
+    {
+        IBlockState curBlockState = world.getBlockState(position);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Local variable 'curBlockState' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 2]
```diff
@@ -568,4 +588,54 @@ public EntityCitizen getCitizen()
...
+
+        if (!crops.isMaxAge(curBlockState))
+        {
+            return false;
+        }
+
+        final ItemStack tool = worker.getHeldItemMainhand();
+
+        //calculate fortune enchantment
+        int fortune = Utils.getFortuneOf(tool);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Local variable 'fortune' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 3]
```diff
@@ -568,4 +588,54 @@ public EntityCitizen getCitizen()
...
+
+    private boolean harvestCrop(final BlockPos position)
+    {
+        IBlockState curBlockState = world.getBlockState(position);
+
+        if (!(curBlockState.getBlock() instanceof IGrowable && curBlockState.getBlock() instanceof BlockCrops))
+        {
+            return false;
+        }
+        BlockCrops crops = (BlockCrops) curBlockState.getBlock();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Local variable 'crops' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 4]
```diff
@@ -568,4 +588,54 @@ public EntityCitizen getCitizen()
...
+
+        //calculate fortune enchantment
+        int fortune = Utils.getFortuneOf(tool);
+
+        final List<ItemStack> drops = crops.getDrops(world, position, curBlockState, fortune);
+
+        final Item seedItem = crops.getItemDropped(curBlockState, world.rand, fortune);
+        if (seedItem != null)
+        {
+            for (Iterator<ItemStack> iterator = drops.iterator(); iterator.hasNext();) {
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Move this left curly brace to the beginning of next line of code. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)
MINOR Local variable 'iterator' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 5]
```diff
@@ -568,4 +588,54 @@ public EntityCitizen getCitizen()
...
+
+        world.setBlockState(position, crops.withAge(0));
+
+        //add the drops to the citizen
+        for (ItemStack item : drops)
+        {
+            InventoryUtils.setStack(worker.getInventoryCitizen(), item);
+        }
+
+        worker.addExperience(0.05);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Assign this magic number 0.05 to a well-named constant, and use the constant instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 6]
```diff
@@ -568,4 +588,54 @@ public EntityCitizen getCitizen()
...
+        final List<ItemStack> drops = crops.getDrops(world, position, curBlockState, fortune);
+
+        final Item seedItem = crops.getItemDropped(curBlockState, world.rand, fortune);
+        if (seedItem != null)
+        {
+            for (Iterator<ItemStack> iterator = drops.iterator(); iterator.hasNext();) {
+                final ItemStack drop = iterator.next();
+
+                // Remove a seed, then break.
+                if (! (drop.getItem() == seedItem) || crops instanceof BlockCarrot || crops instanceof BlockPotato) {
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MAJOR Use the opposite operator ("!=") instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1940)
MINOR Move this left curly brace to the beginning of next line of code. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 10 issues
* MAJOR 2 major
* MINOR 8 minor

Watch the comments in this conversation to review them.

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 7]
```diff
@@ -568,4 +594,55 @@ public EntityCitizen getCitizen()
     {
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Local variable 'curBlockState' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 8]
```diff
@@ -568,4 +594,55 @@ public EntityCitizen getCitizen()
...
+        {
+            for (Iterator<ItemStack> iterator = drops.iterator(); iterator.hasNext();) {
+                final ItemStack drop = iterator.next();
+
+                // Remove a seed, then break.
+                if (! (drop.getItem() == seedItem) || crops instanceof BlockCarrot || crops instanceof BlockPotato) {
+                    iterator.remove();
+                    break;
+                }
+            }
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Local variable 'item' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 9]
```diff
@@ -568,4 +594,55 @@ public EntityCitizen getCitizen()
     {
         return worker;
     }
+
+    private boolean harvestCrop(final BlockPos position)
+    {
+        IBlockState curBlockState = world.getBlockState(position);
+
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Local variable 'crops' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 10]
```diff
@@ -568,4 +594,55 @@ public EntityCitizen getCitizen()
...
+                if (! (drop.getItem() == seedItem) || crops instanceof BlockCarrot || crops instanceof BlockPotato) {
+                    iterator.remove();
+                    break;
+                }
+            }
+        }
+
+        world.setBlockState(position, crops.withAge(0));
+
+        //add the drops to the citizen
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Assign this magic number 0.05 to a well-named constant, and use the constant instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 10 issues
* MAJOR 2 major
* MINOR 8 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [EntityAIWorkFarmer.java#L592](https://github.com/Minecolonies/minecolonies/blob/2ffee0d714ec5b9b502d7f595d46a79dcf67e2b3/src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java#L592): The Cyclomatic Complexity of this method "harvestCrop" is 12 which is greater than 10 authorized. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 11]
```diff
@@ -568,4 +594,55 @@ public EntityCitizen getCitizen()
...
+
+        //calculate fortune enchantment
+        int fortune = Utils.getFortuneOf(tool);
+
+        final List<ItemStack> drops = crops.getDrops(world, position, curBlockState, fortune);
+
+        final Item seedItem = crops.getItemDropped(curBlockState, world.rand, fortune);
+        if (seedItem != null)
+        {
+            for (Iterator<ItemStack> iterator = drops.iterator(); iterator.hasNext();) {
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Move this left curly brace to the beginning of next line of code. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)
MINOR Local variable 'iterator' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 12]
```diff
@@ -138,7 +144,11 @@ private AIState prepareForFarming()
 
         if (currentField.needsWork())
         {
-            switch (currentField.getFieldStage())
+        	if(!checkForHoe()&& canGoPlanting(currentField, building)){
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Move this left curly brace to the beginning of next line of code. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 13]
```diff
@@ -568,4 +594,55 @@ public EntityCitizen getCitizen()
...
+
+        world.setBlockState(position, crops.withAge(0));
+
+        //add the drops to the citizen
+        for (ItemStack item : drops)
+        {
+            InventoryUtils.setStack(worker.getInventoryCitizen(), item);
+        }
+
+        worker.addExperience(0.05);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Assign this magic number 0.05 to a well-named constant, and use the constant instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 14]
```diff
@@ -568,4 +594,55 @@ public EntityCitizen getCitizen()
...
+        final List<ItemStack> drops = crops.getDrops(world, position, curBlockState, fortune);
+
+        final Item seedItem = crops.getItemDropped(curBlockState, world.rand, fortune);
+        if (seedItem != null)
+        {
+            for (Iterator<ItemStack> iterator = drops.iterator(); iterator.hasNext();) {
+                final ItemStack drop = iterator.next();
+
+                // Remove a seed, then break.
+                if (! (drop.getItem() == seedItem) || crops instanceof BlockCarrot || crops instanceof BlockPotato) {
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MAJOR Use the opposite operator ("!=") instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1940)
MINOR Move this left curly brace to the beginning of next line of code. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 15]
```diff
@@ -568,4 +594,55 @@ public EntityCitizen getCitizen()
     {
         return worker;
     }
+
+    private boolean harvestCrop(final BlockPos position)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MAJOR The Cyclomatic Complexity of this method "harvestCrop" is 12 which is greater than 10 authorized. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 16]
```diff
@@ -568,4 +594,55 @@ public EntityCitizen getCitizen()
...
+
+        if (!crops.isMaxAge(curBlockState))
+        {
+            return false;
+        }
+
+        final ItemStack tool = worker.getHeldItemMainhand();
+
+        //calculate fortune enchantment
+        int fortune = Utils.getFortuneOf(tool);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Local variable 'fortune' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 17]
```diff
@@ -568,4 +594,55 @@ public EntityCitizen getCitizen()
...
+                    iterator.remove();
+                    break;
+                }
+            }
+        }
+
+        world.setBlockState(position, crops.withAge(0));
+
+        //add the drops to the citizen
+        for (ItemStack item : drops)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Local variable 'item' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 18]
```diff
@@ -568,4 +594,55 @@ public EntityCitizen getCitizen()
...
+    private boolean harvestCrop(final BlockPos position)
+    {
+        IBlockState curBlockState = world.getBlockState(position);
+
+        if (!(curBlockState.getBlock() instanceof IGrowable && curBlockState.getBlock() instanceof BlockCrops))
+        {
+            return false;
+        }
+        
+        BlockCrops crops = (BlockCrops) curBlockState.getBlock();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Local variable 'crops' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 12 issues
* MAJOR 2 major
* MINOR 10 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [EntityAIWorkFarmer.java](https://github.com/Minecolonies/minecolonies/blob/c401258b1d254f06042a4537e5d7a9767d91d869/src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java): Replace all tab characters in this file by sequences of white-spaces. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00105)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java (hunk 15 above)
Comment (Dadadah) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: This little thing might cause me some problems, I have to loop through the items to get rid of a seed, unless you don't care about that. Let me know @Minecolonies/maintainers.

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 19]
```diff
@@ -568,4 +594,57 @@ public EntityCitizen getCitizen()
...
+
+        //calculate fortune enchantment
+        int fortune = Utils.getFortuneOf(tool);
+
+        final List<ItemStack> drops = crops.getDrops(world, position, curBlockState, fortune);
+
+        final Item seedItem = crops.getItemDropped(curBlockState, world.rand, fortune);
+        if (seedItem != null)
+        {
+            for (Iterator<ItemStack> iterator = drops.iterator(); iterator.hasNext();)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Local variable 'iterator' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 20]
```diff
@@ -568,4 +594,57 @@ public EntityCitizen getCitizen()
...
+    private boolean harvestCrop(final BlockPos position)
+    {
+        IBlockState curBlockState = world.getBlockState(position);
+
+        if (!(curBlockState.getBlock() instanceof IGrowable) || !(curBlockState.getBlock() instanceof BlockCrops))
+        {
+            return false;
+        }
+
+        BlockCrops crops = (BlockCrops) curBlockState.getBlock();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Local variable 'crops' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 8 issues
* MAJOR 1 major
* MINOR 7 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [EntityAIWorkFarmer.java](https://github.com/Minecolonies/minecolonies/blob/e3ad6c970b681133f132134475b72644e157fd14/src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java): Replace all tab characters in this file by sequences of white-spaces. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00105)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 21]
```diff
@@ -210,6 +228,70 @@ private boolean canGoPlanting(@NotNull Field currentField, @NotNull BuildingFarm
...
+    private AIState cycle()
+    {
+        @Nullable BuildingFarmer buildingFarmer = getOwnBuilding();
+
+        if (buildingFarmer == null || buildingFarmer.getCurrentField() == null)
+        {
+            return AIState.PREPARING;
+        }
+
+        @Nullable Field field = buildingFarmer.getCurrentField();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Local variable 'field' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 22]
```diff
@@ -210,6 +228,70 @@ private boolean canGoPlanting(@NotNull Field currentField, @NotNull BuildingFarm
         return true;
     }
 
+    private AIState cycle()
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MAJOR Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)
MAJOR The Cyclomatic Complexity of this method "cycle" is 18 which is greater than 10 authorized. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 23]
```diff
@@ -210,6 +228,70 @@ private boolean canGoPlanting(@NotNull Field currentField, @NotNull BuildingFarm
...
+    {
+        @Nullable BuildingFarmer buildingFarmer = getOwnBuilding();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Local variable 'buildingFarmer' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 24]
```diff
@@ -210,6 +228,70 @@ private boolean canGoPlanting(@NotNull Field currentField, @NotNull BuildingFarm
...
+        @Nullable Field field = buildingFarmer.getCurrentField();
+
+        if (field == null  || checkForHoe())
+        {
+            return AIState.PREPARING;
+        }
+
+        if (workingOffset != null)
+        {
+            BlockPos position = field.getLocation().down().south(workingOffset.getZ()).east(workingOffset.getX());
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Local variable 'position' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 13 issues
* MAJOR 3 major
* MINOR 10 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [EntityAIWorkFarmer.java](https://github.com/Minecolonies/minecolonies/blob/f9c22d61c8506c475085cf931644849407112442/src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java): Replace all tab characters in this file by sequences of white-spaces. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00105)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 25]
```diff
@@ -199,17 +217,95 @@ private boolean canGoPlanting(@NotNull Field currentField, @NotNull BuildingFarm
 
         if (shouldTryToGetSeed)
         {
+            int slot = worker.findFirstSlotInInventoryWith(currentField.getSeed());
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Local variable 'slot' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 14 issues
* MAJOR 3 major
* MINOR 11 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [EntityAIWorkFarmer.java](https://github.com/Minecolonies/minecolonies/blob/68c5a998ac83978d8641820a08dfeb13abf75821/src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java): Replace all tab characters in this file by sequences of white-spaces. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00105)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 26]
```diff
@@ -199,17 +190,89 @@ private boolean canGoPlanting(@NotNull Field currentField, @NotNull BuildingFarm
...
 
+    private AIState cycle()
+    {
+        @Nullable BuildingFarmer buildingFarmer = getOwnBuilding();
+       
+        if (buildingFarmer == null || buildingFarmer.getCurrentField() == null  || checkForHoe())
+        {
+            return AIState.PREPARING;
+        }
+        @Nullable Field field = buildingFarmer.getCurrentField();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Local variable 'field' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 27]
```diff
@@ -199,17 +190,89 @@ private boolean canGoPlanting(@NotNull Field currentField, @NotNull BuildingFarm
...
+
+        if (workingOffset != null)
+        {
+            BlockPos position = field.getLocation().down().south(workingOffset.getZ()).east(workingOffset.getX());
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Local variable 'position' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 28]
```diff
@@ -138,25 +145,9 @@ private AIState prepareForFarming()
...
-                    {
-                        return walkToBlock(currentField.getLocation()) ? AIState.PREPARING : AIState.FARMER_PLANT;
-                    }
-                    break;
-                case PLANTED:
-                    return walkToBlock(currentField.getLocation()) ? AIState.PREPARING : AIState.FARMER_HARVEST;
-                default:
-                    break;
-            }
+        	if(!checkForHoe()&& canGoPlanting(currentField, building)){
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Move this left curly brace to the beginning of next line of code. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 14 issues
* MAJOR 2 major
* MINOR 12 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [EntityAIWorkFarmer.java](https://github.com/Minecolonies/minecolonies/blob/989376d6edb2378e8b62cb8386246861df77b830/src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java): Replace all tab characters in this file by sequences of white-spaces. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00105)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 29]
```diff
@@ -138,24 +141,8 @@ private AIState prepareForFarming()
...
-                    if (canGoPlanting(currentField, building))
-                    {
-                        return walkToBlock(currentField.getLocation()) ? AIState.PREPARING : AIState.FARMER_PLANT;
-                    }
-                    break;
-                case PLANTED:
-                    return walkToBlock(currentField.getLocation()) ? AIState.PREPARING : AIState.FARMER_HARVEST;
-                default:
-                    break;
+            if(!checkForHoe() && canGoPlanting(currentField, building)){
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Move this left curly brace to the beginning of next line of code. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 2 issues
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [Field.java#L82](https://github.com/Minecolonies/minecolonies/blob/56cd352c92f657bfddd2097c5ec83cf573f9ce86/src/main/java/com/minecolonies/entity/ai/citizen/farmer/Field.java#L82): Remove this unused "TAG_STAGE" private field. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1068)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 30]
```diff
@@ -11,7 +12,9 @@
 import com.minecolonies.util.BlockUtils;
 import com.minecolonies.util.InventoryUtils;
 import com.minecolonies.util.Utils;
+import net.minecraft.block.BlockCarrot;
 import net.minecraft.block.BlockCrops;
+import net.minecraft.block.BlockPotato;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Remove this unused import 'net.minecraft.block.BlockPotato'. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 31]
```diff
@@ -22,9 +25,11 @@
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
-import static com.minecolonies.entity.ai.citizen.farmer.Field.FieldStage.*;
 import static com.minecolonies.entity.ai.util.AIState.*;
 
+import java.util.Iterator;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Remove this unused import 'java.util.Iterator'. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 32]
```diff
@@ -11,7 +12,9 @@
 import com.minecolonies.util.BlockUtils;
 import com.minecolonies.util.InventoryUtils;
 import com.minecolonies.util.Utils;
+import net.minecraft.block.BlockCarrot;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MINOR Remove this unused import 'net.minecraft.block.BlockCarrot'. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 3 issues
* MINOR 3 minor

Watch the comments in this conversation to review them.

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 33]
```diff
@@ -199,70 +184,101 @@ private boolean canGoPlanting(@NotNull Field currentField, @NotNull BuildingFarm
...
     {
-        @Nullable BuildingFarmer buildingFarmer = getOwnBuilding();
+        @Nullable final BuildingFarmer buildingFarmer = getOwnBuilding();
 
-        if (buildingFarmer == null || buildingFarmer.getCurrentField() == null)
-        {
-            return AIState.PREPARING;
-        }
-        @Nullable Field field = getOwnBuilding().getCurrentField();
+        if (buildingFarmer == null || checkForHoe()) return AIState.PREPARING;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MAJOR Missing curly brace. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00121)
MINOR At most one statement is allowed per line, but 2 statements were found on this line. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00122)

File: src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 34]
```diff
@@ -199,70 +184,101 @@ private boolean canGoPlanting(@NotNull Field currentField, @NotNull BuildingFarm
...
+
+        @Nullable final Field field = buildingFarmer.getCurrentField();
+
+        if (field == null) return AIState.PREPARING;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: MAJOR Missing curly brace. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00121)
MINOR At most one statement is allowed per line, but 2 statements were found on this line. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00122)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 4 issues
* MAJOR 2 major
* MINOR 2 minor

Watch the comments in this conversation to review them.

**Pull request merged**